pkg/objectio/reader.go

```diff
@@ -176,41 +192,99 @@ func (r *ObjectReader) readFooterAndUnMarshal(ctx context.Context, fileSize, siz
 		Offset: fileSize - size,
 		Size:   size,
 	}
-	err = r.allocData(data.Entries, m)
+	data.Entries[0].Data, err = allocData(data.Entries[0].Size, m)
 	if err != nil {
 		return nil, err
 	}
-	defer r.freeData(data.Entries, m)
+	defer r.freeDataEntries(data.Entries, m)
+	data.Entries[0].ToObject = newToObject
 	err = r.object.fs.Read(ctx, data)
 	if err != nil {
 		return nil, err
 	}
 	footer := &Footer{}
-	err = footer.UnMarshalFooter(data.Entries[0].Data)
+	err = footer.UnMarshalFooter(data.Entries[0].Object.([]byte))
 	if err != nil {
 		return nil, err
 	}
 	return footer, err
 }
 
-func (r *ObjectReader) freeData(Entries []fileservice.IOEntry, m *mpool.MPool) {
+type ToObjectFunc = func(r io.Reader, buf []byte) (any, int64, error)
+
+// newDecompressToObject the decompression function passed to fileservice
+func newDecompressToObject(object []byte, size int64, m *mpool.MPool) ToObjectFunc {
+	return func(read io.Reader, buf []byte) (any, int64, error) {
+		var err error
+		var data []byte
+		data = buf
+		if len(buf) == 0 {
+			data, err = io.ReadAll(read)
+			if err != nil {
+				return nil, 0, err
+			}
+		}
+		object, err = allocData(size, m)
+		if err != nil {
+			return nil, 0, err
+		}
+		object, err = compress.Decompress(data, object, compress.Lz4)
+		if err != nil {
+			freeData(buf, m)
+			return nil, 0, err
+		}
+		return object, int64(len(object)), nil
+	}
+}
+
+// newToObject Because the caller needs to use IOEntry.Object,
+// it needs to be passed to the fileservice ToObject function
+func newToObject(read io.Reader, buf []byte) (any, int64, error) {
+	var err error
+	if len(buf) > 0 {
+		return buf, int64(len(buf)), err
+	}
+	data, err := io.ReadAll(read)
+	return data, int64(len(data)), err
+}
+
+// freeDataEntries free all IOEntry.Data in Entries
+func (r *ObjectReader) freeDataEntries(Entries []fileservice.IOEntry, m *mpool.MPool) {
 	if m != nil {
 		for i := range Entries {
-			if Entries[i].Data != nil {
+			if len(Entries[i].Data) > 0 {
 				m.Free(Entries[i].Data)
 			}
 		}
 	}
 }
 
-func (r *ObjectReader) allocData(Entries []fileservice.IOEntry, m *mpool.MPool) (err error) {
+// freeObjectEntries free all IOEntry.Object in objects
+func (r *ObjectReader) freeObjects(objects [][]byte, m *mpool.MPool) {
 	if m != nil {
-		for i := range Entries {
-			Entries[i].Data, err = m.Alloc(int(Entries[i].Size))
-			if err != nil {
-				return
+		for i := range objects {
+			if len(objects[i]) > 0 {
+				m.Free(objects[i])
 			}
 		}
 	}
-	return nil
+}
+
+func allocData(size int64, m *mpool.MPool) (data []byte, err error) {
+	if m != nil {
```

Comment:
When will we ever use m == nil?    Can we just ALWAYS use mpool?    Mixing and matching them can be very risky.

Reply:
Because there are still some external cases that will not be passed into mpool. For example, when the computing layer reads data, the read data is converted into batch, and the computing layer cannot be free it.